File: 1122-relative-sort-array/counting-sort/python/solution.py
class Solution:
    def relativeSortArray(self, arr1: List[int], arr2: List[int]) -> List[int]:
        max_num = max(arr1)
        count = [0] * (max_num + 1)

        # Count frequency of elements in arr1
        for num in arr1:
            count[num] += 1
        
        res = []
        # Relative Sorting using arr2
        for num in arr2:
            res.extend([num] * count[num])
            count[num] = 0
        
        # Processing the remaining elements in ascending order
        for num in range(max_num + 1):
            res.extend([num] * count[num])

        return res

File: 1122-relative-sort-array/counting-sort/java/solution.java
class Solution {
    public int[] relativeSortArray(int[] arr1, int[] arr2) {
        int maxNum = 0;
        for (int num : arr1) 
            maxNum = Math.max(maxNum, num);
        
        // Count frequency of elements in arr1
        int[] count = new int[maxNum + 1];
        for (int num : arr1) 
            count[num]++;
        
        int[] res = new int[arr1.length];
        int index = 0;
        
        // Relative Sorting using arr2
        for (int num : arr2) {
            while (count[num] > 0) {
                res[index++] = num;
                count[num]--;
            }
        }

        // Processing remaining elements in ascending order
        for (int num = 0; num <= maxNum; num++) {
            while (count[num] > 0) {
                res[index++] = num;
                count[num]--;
            }
        }
        return res;
    }
}
